streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java

```diff
@@ -310,6 +317,25 @@ public void run() {
                     cache.resize(size);
                 }
                 stateConsumer.pollAndUpdate();
+
+                if (fetchDeadline != -1) {
+                    if (fetchDeadline > time.milliseconds()) {
+                        try {
+                            globalConsumerClientInstanceId = globalConsumer.clientInstanceId(Duration.ZERO);
+                            clientInstanceIdFuture.complete(globalConsumerClientInstanceId);
+                            fetchDeadline = -1;
+                        } catch (final TimeoutException swallow) {
+                            // swallow
```

Comment:
So, you've got a TimeoutException but the future is not completed. Surely there comes a point at which the future times out too.

Reply:
Yes, this happens in the `else` of `if (fetchDeadline > time.milliseconds()) {` (from above) further below.

Reply:
Got it.